src/main/kotlin/fi/metatavu/metaform/server/rest/TemplatesApi.kt

```diff
@@ -0,0 +1,139 @@
+
+  @Inject
+  lateinit var logger: Logger
+
+  @Inject
+  lateinit var exportThemeController: ExportThemeController
+
+  @Inject
+  lateinit var templateController: TemplateController
+
+  @Inject
+  lateinit var templateTranslator: TemplateTranslator
+
+  override fun createTemplate(template: Template): Response {
+    val userId = loggedUserId ?: return createForbidden(UNAUTHORIZED)
+
+    if (!isMetatavuAdmin && !isRealmSystemAdmin) {
+      return createForbidden(createNotAllowedMessage(CREATE, TEMPLATE))
+    }
+
+    val templateData = template.data
+            ?: return createBadRequest("Template data is required")
+
+    val templateVisibility = template.visibility
+            ?: return createBadRequest("Template visibility is required")
+
+    val createdTemplate = templateController.createTemplate(
+        templateData = templateData,
+        visibility = templateVisibility,
+        creatorId = userId
+    )
+
+    val templateEntity: Template = try {
+      templateTranslator.translateTemplate(template = createdTemplate)
+    } catch (e: DeserializationFailedException) {
+      return createInternalServerError(e.message)
+    }
+
+    return createOk(templateEntity)
+  }
+
+  override fun deleteTemplate(templateId: UUID): Response {
+    loggedUserId ?: return createForbidden(UNAUTHORIZED)
+
+    if (!isMetatavuAdmin && !isRealmSystemAdmin) {
+      return createForbidden(createNotAllowedMessage(CREATE, TEMPLATE))
+    }
+
+    val template = templateController.findTemplateById(templateId)
+      ?: return createNotFound(createNotFoundMessage("template", templateId))
+
+    templateController.deleteTemplate(template)
+
+    return createNoContent()
+  }
+
+  override fun findTemplate(templateId: UUID): Response {
+    loggedUserId ?: return createForbidden(UNAUTHORIZED)
+
+    if (!isMetatavuAdmin && !isRealmSystemAdmin) {
+      return createForbidden(createNotAllowedMessage(CREATE, TEMPLATE))
+    }
+
+    val template = templateController.findTemplateById(templateId)
+      ?: return createNotFound(createNotFoundMessage("template", templateId))
+
+    return try {
+      createOk(templateTranslator.translateTemplate(template))
+    } catch (e: DeserializationFailedException) {
+      createInternalServerError(e.message)
+    }
+  }
+
+  override fun updateTemplate(templateId: UUID, template: Template): Response {
+    val userId = loggedUserId
+        ?: return createForbidden(UNAUTHORIZED)
+
+    if (!isMetatavuAdmin && !isRealmSystemAdmin) {
+      return createForbidden(createNotAllowedMessage(CREATE, TEMPLATE))
+    }
+
+    val foundTemplate = templateController.findTemplateById(templateId)
+        ?: return createNotFound(createNotFoundMessage(TEMPLATE, templateId))
+
+    val data = try {
+        templateController.serializeTemplateData(templateData = template.data!!)
+    } catch (e: MalformedMetaformJsonException) {
+        createInvalidMessage(createInvalidMessage(TEMPLATE))
+    }
```

Comment:
Template serialization should be moved to controller (template.data should be passed as-is to the controller) and dedicated MalformedTemplateJsonException should be used for detecting malformed JSON